### PR TITLE
[Android] vibration module

### PR DESCRIPTION
**Pull request by skv-headless**
I will fix other notes from https://github.com/facebook/react-native/pull/2794 if I get positive feedback.

### CODE REVIEW

**Comment by nicklockwood** [comment]
Cc: @dmmiller

File: Libraries/Vibration/Vibration.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+var RCTVibration = require('NativeModules').Vibration;
+
+var invariant = require('invariant');
+
+/**
+ * The Vibration API is exposed at `Vibration.vibrate()`. On iOS, calling this
+ * function will trigger a one second vibration. The vibration is asynchronous
+ * so this method will return immediately.
+ *
+ * There will be no effect on devices that do not support Vibration, eg. the simulator.
```
Comment (mkonicek) [inline Libraries/Vibration/Vibration.js]: nit: e.g.

File: Libraries/Vibration/Vibration.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * add `<uses-permission android:name="android.permission.VIBRATE"/>` to `AndroidManifest.xml`
+ *
+ * Vibration patterns are currently unsupported.
+ */
+
+var Vibration = {
+  vibrate: function(duration=400) {
+    invariant(
+      arguments[0] === undefined,
+      'Vibration patterns not supported.'
```
Comment (mkonicek) [inline Libraries/Vibration/Vibration.js]: Why does the message say patterns? Looks like the first argument is `duration`?

File: Libraries/Vibration/Vibration.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * There will be no effect on devices that do not support Vibration, eg. the simulator.
+ *
+ * Note for android
+ * add `<uses-permission android:name="android.permission.VIBRATE"/>` to `AndroidManifest.xml`
+ *
+ * Vibration patterns are currently unsupported.
+ */
+
+var Vibration = {
+  vibrate: function(duration=400) {
```
Comment (mkonicek) [inline Libraries/Vibration/Vibration.js]: Please use Flow to declare it's a number.

File: Libraries/react-native/react-native.js [hunk 4]
```diff
@@ -86,6 +86,7 @@ var ReactNative = {
   get StyleSheet() { return require('StyleSheet'); },
   get TimePickerAndroid() { return require('TimePickerAndroid'); },
   get UIManager() { return require('UIManager'); },
+  get Vibration() { return require('Vibration'); },
```
Comment (mkonicek) [inline Libraries/react-native/react-native.js]: Please also update `react-native.flow.js`. It's a copy of this file for Flow as Flow doesn't understand getters well yet.

**Comment by mkonicek** [comment]
Looks good overall, thanks!

File: Libraries/react-native/react-native.js (hunk 4 above)
Comment (skv-headless) [inline Libraries/react-native/react-native.js]: done

File: Libraries/Vibration/Vibration.js (hunk 1 above)
Comment (skv-headless) [inline Libraries/Vibration/Vibration.js]: sorry didn't understand you :disappointed:

File: Libraries/Vibration/Vibration.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * Vibration patterns are currently unsupported.
+ */
+
+var Vibration = {
+  vibrate: function(duration: number = 400) {
+    if (Platform.OS === 'android') {
+      RCTVibration.vibrate(duration);
+    } else {
+      RCTVibration.vibrate();
```
Comment (skv-headless) [inline Libraries/Vibration/Vibration.js]: it is not that easy to make vibration duration in ios. I would rather make separate pr for that.

Comment (skv-headless) [inline Libraries/Vibration/Vibration.js]: @nicklockwood http://stackoverflow.com/a/13047464/1410905 if it is good solution I can make ios vibration duration here.

File: ReactAndroid/src/main/java/com/facebook/react/modules/vibration/VibrationModule.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+import android.content.Context;
+import android.os.Vibrator;
+
+import com.facebook.react.bridge.ReactApplicationContext;
+import com.facebook.react.bridge.ReactContextBaseJavaModule;
+import com.facebook.react.bridge.ReactMethod;
+
+public class VibrationModule extends ReactContextBaseJavaModule {
+  ReactApplicationContext reactContext;
```
Comment (dmmiller) [inline ReactAndroid/src/main/java/com/facebook/react/modules/vibration/VibrationModule.java]: You don't need to store this separately.  It is available on ReactContextBaseJavaModule via getReactApplicationContext()

File: ReactAndroid/src/main/java/com/facebook/react/modules/vibration/VibrationModule.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
+package com.facebook.react.modules.vibration;
```
Comment (dmmiller) [inline ReactAndroid/src/main/java/com/facebook/react/modules/vibration/VibrationModule.java]: Should add the default license header stuff please.

File: Examples/UIExplorer/UIExplorerList.ios.js [hunk 8]
```diff
@@ -241,6 +241,10 @@ var APIExamples: Array<UIExplorerExample> = [
     module: require('./VibrationIOSExample'),
```
Comment (dmmiller) [inline Examples/UIExplorer/UIExplorerList.ios.js]: Can we get rid of this example as well then?

**Comment by dmmiller** [comment]
Can we also add comments about deprecating VibrationIOS since we will now just have one Vibration API?

**Comment by satya164** [comment]
Since we're adding this, it would be great to support a vibration pattern. Seems pretty simple to implement - http://developer.android.com/reference/android/os/Vibrator.html

File: ReactAndroid/src/main/java/com/facebook/react/modules/vibration/VibrationModule.java (hunk 7 above)
Comment (skv-headless) [inline ReactAndroid/src/main/java/com/facebook/react/modules/vibration/VibrationModule.java]: done

**Comment by skv-headless** [comment]
@dmmiller I've fixed your notes.

**Comment by skv-headless** [comment]
@satya164 I would love to do it.

**Comment by satya164** [comment]
@skv-headless Feel free to do it in a follow-up PR :D

**Comment by skv-headless** [comment]
I think this PR is finished. I've already started to work on vibration patters but I'm going to create separate PR for that.

**Comment by dmmiller** [comment]
@facebook-github-bot shipit

**Comment by dmmiller** [comment]
Cool, thanks.  I've asked for the bot to accept it :)

**Comment by bestander** [comment]
@facebook-github-bot shipit

**Comment by mkonicek** [comment]
Failed to land because of failing tests. Luckily this should be easy, you just need to update BUCK files so that the code builds internally at fb where we use Buck.

See https://circleci.com/gh/facebook/react-native/3133

My guess is you just need to add a dependency here: https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/shell/BUCK

**Comment by skv-headless** [comment]
@mkonicek is there any instruction how to launch tests locally? I'll update tomorrow.

**Comment by skv-headless** [comment]
@mkonicek circle tests passed

**Comment by dmmiller** [comment]
@facebook-github-bot shipit

**Comment by skv-headless** [comment]
@dmmiller if there are problems with shipping (maybe like this https://github.com/facebook/react-native/pull/6208#issuecomment-190266569) I can create PR again

**Comment by dmmiller** [comment]
I'll get it to land.  Let me commandeer the diff locally and get it in.

**Comment by dmmiller** [comment]
@facebook-github-bot shipit
